release/k8s_tests/run_gcs_ft_on_k8s.py

```diff
@@ -1,10 +1,323 @@
-from kubernetes import client, config
+import subprocess
+from kubernetes import client, config, watch
+import requests
+import random
+import uuid
+import pathlib
+import time
+import ray
+import os
+
+# global variables for the cluster informations
+cluster_id = str(uuid.uuid4()).split("-")[0]
+ray_cluster_name = "cluster-" + cluster_id
+ray_service_name = "service-" + cluster_id
+locust_id = "ray-locust-" + cluster_id
+
+
+if os.environ.get("RAY_IMAGE") is not None:
+    ray_image = os.environ.get("RAY_IMAGE")
+elif ray.__commit__ == "{{RAY_COMMIT_SHA}}":
+    ray_image = "rayproject/ray:nightly"
+else:
+    ray_image = f"rayproject/ray:{ray.__commit__[:6]}"
 
-# Configs can be set in Configuration class directly or using helper utility
 config.load_kube_config()
+cli = client.CoreV1Api()
+
+yaml_path = pathlib.Path("/tmp/ray_v1alpha1_rayservice.yaml")
+
+
+def check_kuberay_installed():
+    # Make sure the ray namespace exists
+    pods = subprocess.check_output(
+        ["kubectl", "get", "pods", "--namespace", "ray-system", "--no-headers"]
+    ).decode()
+    pod_name = pods.split()[0]
+    assert pod_name.startswith("kuberay-operator"), "KubeRay has not been installed."
+
+
+def start_rayservice():
+    # step-1: generate the yaml file
+    print(f"Using ray image: {ray_image}")
+    template = (
+        pathlib.Path("ray_v1alpha1_rayservice_template.yaml")
+        .read_text()
+        .format(cluster_id=cluster_id, ray_image=ray_image)
+    )
+
+    print("=== YamlFile ===")
+    print(template)
+    tmp_yaml = pathlib.Path("/tmp/ray_v1alpha1_rayservice.yaml")
+    tmp_yaml.write_text(template)
+
+    print("=== Get Pods from ray-system ===")
+    print(
+        subprocess.check_output(
+            ["kubectl", "get", "pods", "--namespace", "ray-system", "--no-headers"]
+        ).decode()
+    )
+
+    # step-2: create the cluter
+    print(f"Creating cluster with id: {cluster_id}")
+    print(subprocess.check_output(["kubectl", "create", "-f", str(tmp_yaml)]).decode())
+
+    # step-3: make sure the ray cluster is up
+    w = watch.Watch()
+    start_time = time.time()
+    head_pod_name = None
+    for event in w.stream(
+        func=cli.list_namespaced_pod,
+        namespace="default",
+        label_selector=f"rayCluster={ray_cluster_name},ray.io/node-type=head",
+        timeout_seconds=60,
```

Comment:
It might take longer for the head pod to start running if the Ray container needs to be pulled.

Nice use of the k8s watch functionality! (First thing that would come to my head would be writing a loop that polls.)

I wrote some test utilities here:
https://github.com/ray-project/ray/blob/5bf6562f384fad8784398cad3a9eddd29c89a51a/python/ray/tests/kuberay/utils.py
May or may not be convenient to use those here, but just FYI :)

Reply:
Nice! I'll check your toolbox and use that here! I wish I knew it earlier.